onnx/defs/tensor/defs.cc

```diff
@@ -77,19 +77,126 @@ could also be 0, in which case the actual dimension value is unchanged (i.e. tak
+          // The targetShape vector represents the specified shape for output.
+          std::vector<int64_t> targetShape;
+          if (targetShapeInitializer->has_raw_data()) {
+            const std::string &bytes = targetShapeInitializer->raw_data();
+            targetShape.insert(targetShape.end(),
+                               reinterpret_cast<const int64_t *>(bytes.c_str()),
+                               reinterpret_cast<const int64_t *>(bytes.c_str() +
+                                                                 bytes.size()));
+          } else {
+            const auto &data = targetShapeInitializer->int64_data();
+            targetShape.insert(targetShape.end(), data.begin(), data.end());
+          }
+
+          // Iterate through targetShape, adding dimensions in the outputShape
+          // TensorProto. If the targertShape dimension is -1, we do not set the
+          // dimension value in this iteration, but we record the Dimension. If
+          // targertShape dimension is 0, we attempt to propagate the dimension
+          // value/param. If the value cannot be inferred, we set the flag in
+          // the unresolveZeros vector. If targetShape dimension is positive, we
+          // set the dimension value in the outputShape. We track the product of
+          // the dimensions we are setting outputShape in the outputProduct
+          // variable. The outputProduct will potentially be used for inferring
+          // a dimension marked -1.
+          auto *outputShape =
+              ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          TensorShapeProto::Dimension *negativeOneDim = nullptr;
+          const auto &dataInputTensorType = ctx.getInputType(0)->tensor_type();
+          std::vector<bool> unresolvedZeros(targetShape.size(), false);
+          int64_t outputProduct = 1;
+          for (int i = 0; i < static_cast<int>(targetShape.size()); ++i) {
+            // Add a new dimension to outputShape
+            auto *new_dim = outputShape->add_dim();
+            if (targetShape[i] == -1) {
+              // Check if multiple -1's. If not, set negativeOneDim, marking
+              // this dimension to potentially be filled in later.
+              if (negativeOneDim) {
+                fail_shape_inference(
+                    "Target shape may not have multiple -1 dimensions");
+              }
+              negativeOneDim = new_dim;
+            } else if (targetShape[i] == 0) {
+              // Check if data input has a shape and if the index i is within
+              // its bounds. If these conditions are satisfied, any dimension
+              // value/param should be propogated. If dimension value cannot be
+              // inferred, set the corresponding  unresolvedZeros flag to true.
+              unresolvedZeros[i] = true;
+              if (dataInputTensorType.has_shape() &&
+                  i < dataInputTensorType.shape().dim_size()) {
```

Comment:
if (i >= dataInputTensorType.shape().dim_size()), there should be a data error, right?

Reply:
Yes, good point - I'll add a fail_inference for that